prov/efa/src/rdm/efa_rdm_cq.c

```diff
@@ -138,111 +138,239 @@ void efa_rdm_cq_proc_ibv_recv_rdma_with_imm_completion(
 	}
 }
 
-#if HAVE_EFADV_CQ_EX
 /**
- * @brief Read peer raw address from EFA device and look up the peer address in AV.
- * This function should only be called if the peer AH is unknown.
- * @return Peer address, or FI_ADDR_NOTAVAIL if unavailable.
+ * @brief Populate the efa_ep_addr struct with the raw address, QPN and QKEY
+ * This function tries to read the raw address first from the packet header
+ * If the packet header does not have the raw address, it uses efadv_wc_read_sgid
+ * if available.
+ * @return FI_SUCCESS if the efa_ep_addr struct is populated. Error otherwise.
  */
-static inline
-fi_addr_t efa_rdm_cq_determine_peer_address_from_efadv(
-						       struct ibv_cq_ex *ibv_cqx,
-						       enum ibv_cq_ex_type ibv_cq_ex_type)
+static inline int
+efa_rdm_cq_populate_src_efa_ep_addr(struct efa_rdm_pke *pkt_entry, struct ibv_cq_ex *ibv_cqx,
+					     enum ibv_cq_ex_type ibv_cq_ex_type, struct efa_ep_addr *efa_ep_addr)
 {
-	struct efa_rdm_pke *pkt_entry;
+	struct efa_rdm_base_hdr *base_hdr;
 	struct efa_rdm_ep *ep;
-	struct efa_ep_addr efa_ep_addr = {0};
-	fi_addr_t addr;
-	union ibv_gid gid = {0};
-	uint32_t *connid = NULL;
+	void *raw_addr_ptr = NULL;
+	uint32_t *connid;
+
+	ep = pkt_entry->ep;
+	assert(ep);
+
```

Comment:
We missed a check here earlier. It needs to check whether the `pkt_entry` has hdr or not. For zcpy recv (`if (pkt_entry->alloc_type == EFA_RDM_PKE_FROM_USER_RX_POOL)`), the `pkt_entry` has no hdr. In that case you cannot get either raw addr or connid from the hdr. Though I never saw a zcpy rx pkt can run into this corner case. Because zcpy rx pkt can only be kicked off after the sender got a handshake from the receiver.

Reply:
That case should be covered by this check
https://github.com/sunkuamzn/libfabric/blob/5583625cf37a43444785c70872d0ea0ab9a99ba6/prov/efa/src/rdm/efa_rdm_cq.c#L186

`efa_rdm_pke_get_req_raw_addr` returns NULL if the `EFA_RDM_REQ_OPT_RAW_ADDR_HDR` flag is not set and the raw address is not present

Reply:
I see what you mean. Fixed